process_settings.go

```diff
@@ -0,0 +1,199 @@
+// Package process_settings implements dynamic settings for the process.
+package process_settings
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"reflect"
+	"strings"
+
+	"github.com/fsnotify/fsnotify"
+	"gopkg.in/yaml.v3"
+)
+
+// A ProcessSettings is a collection of settings files and a target evaluator
+// that can be used to get the value of a settings based on the current targeting.
+type ProcessSettings struct {
+	FilePath            string            // The path to the settings file that was used to create the ProcessSettings
+	Settings            *[]SettingsFile   // The settings files that make up the ProcessSettings
+	TargetEvaluator     TargetEvaluator   // The target evaluator that is used to determine which settings files are applicable
+	Monitor             *fsnotify.Watcher // The file monitor that is used to detect changes to the settings file
+	WhenUpdatedRegistry []func()          // A list of functions to call when the settings are updated
+}
+
+type SettingNotFound struct {
+	NotFoundKey string
+}
+
+// NewProcessSettingsFromFile creates a new instance of ProcessSettings by
+// loading the settings from a specified file path and using the specified
+// static context to evaluate the targeting.
+func NewProcessSettingsFromFile(filePath string, staticContext map[string]interface{}) (*ProcessSettings, error) {
+	settings, err := loadSettingsFromFile(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	monitor, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	err = monitor.Add(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ProcessSettings{
+		FilePath:        filePath,
+		Settings:        settings,
+		TargetEvaluator: TargetEvaluator{staticContext},
+		Monitor:         monitor,
+	}, nil
+}
+
+// Get returns the value of a setting based on the current targeting.
+// If the requested setting is not found, an error is returned.
+func (ps *ProcessSettings) Get(settingPath ...string) (interface{}, error) {
+	var value interface{}
+
+	valueFound := false
+	for _, settingsFile := range *ps.Settings {
+		if ps.TargetEvaluator.isTargetMatch(settingsFile) {
+			if fileValue, keyExists := dig(settingsFile.Settings, settingPath...); keyExists {
+				value = fileValue
+				valueFound = true
+			}
+		}
+	}
+
+	if !valueFound {
+		return nil, errors.New(fmt.Sprintf("The setting '%s' was not found", dotDelimitedSettingsPath(settingPath)))
+	}
+
+	return value, nil
+}
+
+// SafeGet returns the value of a setting based on the current targeting.
+// If the requested setting is not found, nil is returned.
+func (ps *ProcessSettings) SafeGet(settingPath ...string) interface{} {
+	value, _ := ps.Get(settingPath...)
```

Comment:
I think this should check the error for `The setting '%s' was not found` and return `nil` just in that case. Otherwise it should return the `err`.

Reply:
Doesn't that defeat the point of `SafeGet` if it can also return an error? It's supposed to be "safe" and unable to return errors that require acting upon. I would say that we could log the error returned here, but also by `Get`'s contract  the only error returned is if the value is not found, so I'm not sure why we need to do the extra verification here.

Reply:
It's mimicking `optional: true`, right? So I think you should make this the bottom method:
```
func (ps *ProcessSettings) OptionalGet(settingPath ...string) 
```
and have it return `nil` if unset (as well as any errors it encounters while doing that). Then layer this above that
```
func (ps *ProcessSettings) Get(settingPath ...string)
```
which will return any of the errors from `OptionalGet` but also the error that the setting isn't set.

Reply:
Went ahead and implemented the error handling using the pattern we discussed live de4d14abe844a3baad5f4e5b59f0a766fe952386

Reply:
Went ahead and implemented the error handling using the pattern we discussed live de4d14abe844a3baad5f4e5b59f0a766fe952386